src/examples/libpmem2/redo.c

```diff
@@ -0,0 +1,388 @@
+	uint64_t head;
+	uint64_t pool_size;
+};
+
+struct pool_layout {
+	struct pool_hdr hdr;
+	struct node nodes[];
+};
+
+/*
+ * redo_apply -- process and apply the redo log
+ */
+static void
+redo_apply(struct pool_layout *pool)
+{
+	struct redo_log *redo = &pool->hdr.redo;
+
+	if (!redo->apply || !redo->last) {
+		if (redo->apply || redo->last) {
+			redo->last = 0;
+			redo->apply = 0;
+			Persist(&redo, 256); /* persist entire CL */
+		}
+		return; /* no redo log to process */
+	}
+
+	uint8_t *start = (uint8_t *)pool;
+	for (int i = 0; i < redo->last; ++i) {
+		uint64_t *node = (uint64_t *)&start[redo->entries[i].offset];
+		*node = redo->entries[i].data;
+		Flush(node, sizeof(*node));
+	}
+
+	Drain();
+	redo->last = 0;
+	redo->apply = 0;
+	Persist(&redo, 256); /* persist entire CL */
+}
+
+/*
+ * redo_add -- add an entry to redo log
+ */
+static void
+redo_add(struct redo_log *redo, uintptr_t offset, uint64_t data)
+{
+	assert(redo->apply == 0);
+	if (redo->last == REDO_LOG_SIZE) {
+		fprintf(stderr, "redo log to small");
+		exit(1);
+	}
+
+	struct redo_log_entry *entry = &redo->entries[redo->last++];
+
+	entry->offset = (uintptr_t)offset;
+	entry->data = data;
+	/*
+	 * To achive optimal performance we should flush only
+	 * full cachelines if it is possible
+	 */
+	if (redo->last % 4 == 0)
+		Flush(entry - 3, sizeof(*entry) * 4);
+}
+
+/*
+ * redo_commit -- commit redo log
+ */
+static void
+redo_commit(struct redo_log *redo)
+{
+	if (redo->last == 0)
+		return;
+
+	/* flush last not full cacheline if needed */
+	int left = redo->last % 4;
+	if (left != 0)
+		Flush(&redo->entries[redo->last] - (left - 1),
+			sizeof(*redo->entries) * left);
+
+	/* drain all previous flushes */
+	Drain();
+	redo->apply = 1;
+	Persist(&redo->apply, sizeof(redo->last));
+}
+
+/*
+ * list_add -- add a new node to the list
+ */
+static void
+list_add(struct pool_layout *pool, uint64_t key, uint64_t value)
+{
+	struct node *node;
+	struct node *next = NULL;
+	struct node *prev = NULL;
+
+	if (pool->hdr.pool_size < pool->hdr.free) {
+		fprintf(stderr, "pool is full\n");
+		exit(1);
+	}
+
+	node = &pool->nodes[pool->hdr.free];
+	if (pool->hdr.free != 0) {
+		next = &pool->nodes[pool->hdr.head];
+		while (next->key < key) {
+			if (next->next == UINT64_MAX) {
+				prev = next;
+				next = NULL;
+				break;
+			}
+			prev = next;
+			next = &pool->nodes[next->next];
+		}
+	}
+	struct redo_log *redo = &pool->hdr.redo;
+	if (next == NULL) {
+		redo_add(redo, offset(pool, &node->next), UINT64_MAX);
+	} else {
+		redo_add(redo, offset(pool, &node->next), next->this);
+		redo_add(redo, offset(pool, &next->prev), pool->hdr.free);
+	}
+
+	if (prev == NULL) {
+		redo_add(redo, offset(pool, &node->prev), UINT64_MAX);
+		redo_add(redo, offset(pool, &pool->hdr.head), pool->hdr.free);
+	} else {
+		redo_add(redo, offset(pool, &node->prev), prev->this);
+		redo_add(redo, offset(pool, &prev->next), pool->hdr.free);
+	}
+
+	redo_add(redo, offset(pool, &node->key), key);
```

Comment:
this is a newly "allocated" node, it doesn't need to be updated through a redo log.